ouroboros-consensus/test-consensus/Test/Dynamic/Util.hs

```diff
@@ -221,34 +233,219 @@ leaderScheduleFromTrace (NumSlots numSlots) = LeaderSchedule .
+        -- introduce an additional competing (e.g. long enough) chain.
+        --
+        -- This can happen even if an older node is also leading in this slot
+        -- (ie lo /= 0). It is possible for such an older node to send all of
+        -- its /old/ blocks to a joining leader before they both then forge
+        -- distinct new blocks atop that old chain.
+        --
+        -- (This has only been observed so far when the pre-existing chains
+        -- were of length 1, lo = 0, and hi = 1. See e.g.
+        -- https://hydra.iohk.io/build/991294)
+      Just n
+        | n > 1                 -> extendingTheRun
+        | n < 1                 -> (longest, current)
+        | otherwise {- n = 1 -} -> (longest, noRun)
+      where
+        -- INVARIANT: 0 <= lo <= (n = number of actual leaders) <= hi
+        nCertain = if lo == hi then Just lo else Nothing
+        lo = numberLeadersOld
+        hi = numberLeadersOld + numberLeadersNew
+
+        -- count only scheduled leaders that have already joined the network
+        numberLeadersOld = count (<)
+        -- count only scheduled leaders that are currently joining the network
+        numberLeadersNew = count (==)
+
+        count rel = length $ filter keep leaders
+          where
+            keep (CoreNodeId nid) = joinSlots !! nid `rel` slot
+
+        extendingTheRun = (Just longest', Just current')
+          where
+            current' = incCrowdedRun current slot
+            longest' = case longest of
+                Just old
+                  | crowdedRunLength new > crowdedRunLength old -> new
+                  | otherwise -> old
+                Nothing -> new
+              where
+                new = current'
 
-noRun :: CrowdedRun
-noRun = CrowdedRun Nothing
+{-------------------------------------------------------------------------------
+  Node Join Plan - when each node joins the test network
+-------------------------------------------------------------------------------}
 
-incCrowdedRun :: SlotNo -> CrowdedRun -> CrowdedRun
-incCrowdedRun slot (CrowdedRun Nothing)          = CrowdedRun (Just (slot, 1))
-incCrowdedRun _    (CrowdedRun (Just (slot, n))) = CrowdedRun (Just (slot, n + 1))
+-- | A list of durations as number of slots. Each duration is the amount of
+-- time after a node joins the network before either the next node does or the
+-- test ends.
+--
+-- ASSUMPTIONS:
+-- * Nodes join in the order of their Id.
+-- * Node 0 joins first, at slot 0, so the first duration in the list occurs
+-- after that.
+-- * Nodes other than Node 0 join after some possibly empty duration
+-- since the previous node joined.
+-- * The last duration in the list is the run duration after the last node
+-- joins.
+-- * The number of durations is therefore the number of nodes.
+--
+-- This list is a \"weak composition\" in terms of
+-- <https://en.wikipedia.org/wiki/Composition_(combinatorics)> for the total
+-- plan duration.
+newtype NodeJoinPlan = NodeJoinPlan [Int]
```

Comment:
Hmm, I'm still thinking that we should not overload the last `Int` to mean the final duration, so something like:
```haskell
data NodeJoinPlan = NodeJoinPlan
  { joinIntervals :: [Int]
    -- ^ TODO
  , finalDuration :: Int
    -- ^ TODO
  }
```
With better names and possibly an unsigned type (we often use `Word64`, even though we don't need the full size).

I don't think this is harder to generate or shrink. In fact, I think this type should be easier in use in general, as we're not using `last` and `++ [..]` the whole time.

What do you think?

Reply:
Will do.